src/topology/config/watcher.rs

```diff
@@ -0,0 +1,125 @@
+use super::Config;
+use notify::{watcher, DebouncedEvent, RecommendedWatcher, RecursiveMode, Watcher};
+use std::{
+    path::{Path, PathBuf},
+    sync::mpsc::{channel, Receiver},
+    thread,
+    time::Duration,
+};
+
+/// Per notify own documentation, it's advised to have delay of more than 30 sec,
+/// so to avoid receiving repetitions of previous events on macOS.
+///
+/// But, config and topology reload logic can handle:
+///  - Invalid config, caused either by user or by data race.
+///  - Frequent changes, caused by user/editor modifying/saving file in small chunks.
+/// so we can use smaller, more responsive delay.
+pub const CONFIG_WATCH_DELAY: std::time::Duration = std::time::Duration::from_secs(1);
+
+const RETRY_TIMEOUT: std::time::Duration = std::time::Duration::from_secs(10);
+
+/// On unix triggers SIGHUP when file on config_path changes.
+/// Accumulates file changes until no change for given duration has occured.
+/// Has best effort guarante of detecting all file changes from the layend of
+/// this function until the main thread stops.
+///
+/// Doesn't do anything on Windows.
+pub fn config_watcher(config: &Config, config_path: PathBuf, delay: Duration) {
+    if config.global.reload_config {
+        #[cfg(unix)]
+        {
+            use nix::sys::signal;
+            // Create watcher now so not to miss any changes happening between
+            // returning from this function and thread started.
+            let mut watcher = create_watcher(&config_path, delay);
+
+            thread::spawn(move || loop {
+                if let Some((_, receiver)) = watcher.take() {
+                    info!("Watching configuration file");
+                    while let Ok(msg) = receiver.recv() {
+                        match msg {
+                            DebouncedEvent::Write(_)
+                            | DebouncedEvent::Create(_)
+                            | DebouncedEvent::Remove(_) => {
+                                info!("Configuration file changed");
+                                let _ = signal::raise(signal::Signal::SIGHUP).map_err(|error| {
+                                    debug!(message = "Unable to raise SIGHUP.", ?error)
```

Comment:
I think this makes more sense to be an error, what do you think?

```suggestion
                                    error!(message = "Unable to raise SIGHUP.", ?error)
```

Reply:
It does, as now it ocured to me that user can do something about resolving the error. That is to just reload Vector. So I'll also change the message to reflect that.